File: frontend/src/lib/matching/explanations.ts
/** Template explanations — mirror backend/app/explanations.py (no LLM). */

import type { LocalizedText, MatchProfile, SchemeRecord } from "./types";

export const DISCLAIMER_EN =
  "Based on published eligibility rules only; confirm with the implementing office. Not legal advice.";
export const DISCLAIMER_ML =
  "പ്രസിദ്ധീകരിച്ച യോഗ്യതാ നിബന്ധനകളെ അടിസ്ഥാനമാക്കി മാത്രം; നടപ്പാക്കുന്ന ഓഫീസുമായി സ്ഥിരീകരിക്കുക. നിയമ ഉപദേശമല്ല.";

function normCountry(country: string | null | undefined): string {
  return (country || "India").trim().toLowerCase().replace(/-/g, "_").replace(/ /g, "_") || "india";
}

const COUNTRY_DISPLAY: Record<string, string> = {
  india: "India",
  united_states: "United States",
  usa: "United States",
  us: "United States",
  united_kingdom: "United Kingdom",
  uk: "United Kingdom",
  gb: "United Kingdom",
  great_britain: "United Kingdom",
  canada: "Canada",
  can: "Canada",
};

/** Human-readable country name for explanations (aliases → canonical display name). */
export function countryDisplay(country: string | null | undefined): string {
  return COUNTRY_DISPLAY[normCountry(country)] || (country || "India").trim();
}

function isIndia(country: string | null | undefined): boolean {
  return normCountry(country) === "india";
}

function fmtIncome(value: number | null | undefined, country?: string | null): string {
  if (value == null) return "n/a";
  const c = normCountry(country);
  const n = Math.trunc(value);
  if (c === "united_states" || c === "usa" || c === "us") return `$${n.toLocaleString("en-US")}`;
  if (c === "united_kingdom" || c === "uk" || c === "gb" || c === "great_britain") {
    return `£${n.toLocaleString("en-GB")}`;
  }
  if (c === "canada" || c === "can") return `C$${n.toLocaleString("en-CA")}`; // CAD, never bare "$"
  return `Rs.${n.toLocaleString("en-IN")}`;
}

function impliesLowGateLabel(country: string | null | undefined): string {
  const c = (country || "India").trim().toLowerCase().replace(/-/g, "_").replace(/ /g, "_") || "india";
  if (c === "united_states" || c === "usa" || c === "us") return "$60,000"; // soft US heuristic — not official FPL
  if (c === "united_kingdom" || c === "uk" || c === "gb" || c === "great_britain") return "£60,000"; // soft UK heuristic (HICBC anchor)
  if (c === "canada" || c === "can") return "C$58,523"; // soft Canada heuristic (lowest federal bracket / CLB line)
  if (c === "india" || c === "") return "Rs.5,00,000";
  return "n/a (no soft gate for this country)";
}


function rulePhraseEn(rule: string, scheme: SchemeRecord, profile: MatchProfile): string {
  const rules = (scheme.eligibility_rules || {}) as Record<string, unknown>;
  switch (rule) {
    case "min_age":
      return `age ${profile.age} meets minimum age ${rules.min_age}`;
    case "max_age":
      return `age ${profile.age} within maximum age ${rules.max_age}`;
    case "max_annual_income":
      return `annual income ${fmtIncome(profile.annual_income, profile.country)} within cap ${fmtIncome(rules.max_annual_income as number, profile.country)}`;
    case "max_monthly_household_income":
      return `monthly household income ${fmtIncome(profile.monthly_household_income, profile.country)} within cap ${fmtIncome(rules.max_monthly_household_income as number, profile.country)}`;
    case "countries":
      return `available in ${countryDisplay(profile.country)}`;
    case "implies_low_income":
      // "BPL/destitute" is Indian terminology — neutral wording elsewhere.
      return `annual income ${fmtIncome(profile.annual_income, profile.country)} within soft low-income gate (${isIndia(profile.country) ? "BPL/destitute" : "low-income / means-tested"}; gate ${impliesLowGateLabel(profile.country)})`;
    case "gender":
      return `gender '${profile.gender}' matches required '${rules.gender}'`;
    case "marital_status":
      return `marital status '${profile.marital_status}' is in allowed list`;
    case "deserted_7_years_over_50":
      return `deserted case with age ${profile.age} (≥50 required)`;
    case "disability_required":
      return "disability requirement satisfied";
    case "min_disability_percent":
      return `disability ${profile.disability_percent}% meets minimum ${rules.min_disability_percent}%`;
    case "occupations":
      return "occupation intersects scheme occupation list";
    case "categories":
      return "category intersects scheme category list";
    case "states":
      return `state '${profile.state || "Kerala"}' is in scheme states`;
    case "land_ownership":
      return "land ownership requirement satisfied";
    case "land_ownership_landless":
      return "profile indicates landless / no ownership";
    case "maternity_required":
      return "maternity / recent delivery signal present (pregnant, lactating, or child ≤6 months)";
    case "primary_breadwinner_deceased_required":
      return "primary breadwinner deceased flag set";
    case "kawwf_member_required":
      return "KAWWF membership requirement satisfied";
    case "min_agri_labour_years":
      return `agri labour years ${profile.agri_labour_years} meets minimum ${rules.min_agri_labour_years}`;
    case "districts":
      return `district '${profile.district}' is in scheme district list`;
    default:
      return rule.replace(/_/g, " ");
  }
}

function rulePhraseMl(rule: string, scheme: SchemeRecord, profile: MatchProfile): string {
  const rules = (scheme.eligibility_rules || {}) as Record<string, unknown>;
  switch (rule) {
    case "min_age":
      return `പ്രായം ${profile.age} ഏറ്റവും കുറഞ്ഞ പ്രായം ${rules.min_age} നിറവേറ്റുന്നു`;
    case "max_annual_income":
      return `വാർഷിക വരുമാനം ${fmtIncome(profile.annual_income, profile.country)} പരിധി ${fmtIncome(rules.max_annual_income as number, profile.country)} യിൽ ഉൾപ്പെടുന്നു`;
    case "max_monthly_household_income":
      return `മാസ വരുമാനം ${fmtIncome(profile.monthly_household_income, profile.country)} പരിധിക്കുള്ളിൽ`;
    case "countries":
      return `${countryDisplay(profile.country)} ൽ ലഭ്യമാണ്`;
    case "implies_low_income":
      return `വാർഷിക വരുമാനം ${fmtIncome(profile.annual_income, profile.country)} താഴ്ന്ന വരുമാന സോഫ്റ്റ് ഗേറ്റിനുള്ളിൽ (${isIndia(profile.country) ? "BPL/destitute" : "താഴ്ന്ന വരുമാനം / വരുമാന പരിശോധന"}; ${impliesLowGateLabel(profile.country)})`;
    case "gender":
      return `ലിംഗം '${profile.gender}' യോജിക്കുന്നു`;
    case "marital_status":
      return `വൈവാഹിക നില '${profile.marital_status}' അനുവദനീയമാണ്`;
    case "disability_required":
      return "ഭിന്നശേഷി നിബന്ധന നിറവേറ്റുന്നു";
    case "min_disability_percent":
      return `ഭിന്നശേഷി ${profile.disability_percent}% കുറഞ്ഞത് ${rules.min_disability_percent}% നിറവേറ്റുന്നു`;
    case "occupations":
      return "തൊഴിൽ പട്ടികയുമായി യോജിക്കുന്നു";
    case "categories":
      return "വിഭാഗം പട്ടികയുമായി യോജിക്കുന്നു";
    case "states":
      return `സംസ്ഥാനം '${profile.state || "Kerala"}' യോജിക്കുന്നു`;
    case "land_ownership":
      return "ഭൂമി ഉടമസ്ഥത നിബന്ധന നിറവേറ്റുന്നു";
    case "maternity_required":
      return "ഗർഭം / മുലയൂട്ടൽ / 6 മാസത്തിനുള്ളിൽ കുട്ടി എന്ന സിഗ്നൽ ഉണ്ട്";
    case "primary_breadwinner_deceased_required":
      return "പ്രധാന വരുമാനദാതാവ് മരിച്ചുവെന്ന ഫ്ലാഗ് സജ്ജമാണ്";
    default:
      return rule.replace(/_/g, " ") + " [ML]";
  }
}

export function buildExplanation(args: {
  scheme: SchemeRecord;
  profile: MatchProfile;
  matched_rules: string[];
  unmatched_rules: string[];
  missing_fields: string[];
  status: string;
}): LocalizedText {
  const { scheme, profile, matched_rules, unmatched_rules, missing_fields, status } = args;
  const name_en = (scheme.scheme_name || {}).en || scheme.id || "scheme";
  const name_ml = (scheme.scheme_name || {}).ml || name_en;

  let body_en: string;
  let body_ml: string;
  if (matched_rules.length) {
    body_en = matched_rules.map((r) => rulePhraseEn(r, scheme, profile)).join("; ") + ".";
    body_ml = matched_rules.map((r) => rulePhraseMl(r, scheme, profile)).join("; ") + ".";
  } else {
    body_en = "No structured hard-filter fields constrained this match.";
    body_ml = "ഘടനാപരമായ കഠിന ഫിൽട്ടറുകൾ ഈ പൊരുത്തത്തെ നിയന്ത്രിച്ചില്ല.";
  }

  const prefix_en =
    status === "uncertain"
      ? `Possibly relevant for ${name_en} (needs verification). `
      : `Likely eligible for ${name_en}: `;
  const prefix_ml =
    status === "uncertain"
      ? `${name_ml} — സ്ഥിരീകരണം ആവശ്യം. `
      : `${name_ml} യ്ക്ക് യോഗ്യതയുണ്ടാകാം: `;

  let extra_en = "";
  let extra_ml = "";
  if (missing_fields.length) {
    extra_en += ` Missing profile fields: ${missing_fields.join(", ")}.`;
    extra_ml += ` കുറവുള്ള ഫീൽഡുകൾ: ${missing_fields.join(", ")}.`;
  }
  if (unmatched_rules.length) {
    extra_en += ` Unmatched: ${unmatched_rules.join(", ")}.`;
  }
  if (profile.district) {
    extra_en += ` District on profile: ${profile.district}.`;
    extra_ml += ` ജില്ല: ${profile.district}.`;
  }
  const rules = (scheme.eligibility_rules || {}) as Record<string, unknown>;
  if (rules.verify) {
    const notes = String(rules.verify_notes || "");
    extra_en += " Official rules marked verify=true";
    if (notes) {
      extra_en += ` (${notes.slice(0, 180)}${notes.length > 180 ? "…" : ""})`;
    }
    extra_en += ".";
    extra_ml += " ഔദ്യോഗിക നിബന്ധനകൾ verify=true ആയി അടയാളപ്പെടുത്തിയിരിക്കുന്നു.";
  }

  const en_text = (prefix_en + body_en + extra_en + " " + DISCLAIMER_EN).trim();
  return {
    en: en_text,
    ml: (prefix_ml + body_ml + extra_ml + " " + DISCLAIMER_ML).trim(),
    // Hindi scheme-body explanations fall back to English pending native review.
    hi: en_text,
  };
}

File: backend/app/explanations.py
"""Template explanations from matched rules. Optional LLM stub (never decides eligibility)."""

from __future__ import annotations

import os
from typing import Any

from .models import LocalizedText, MatchProfile


DISCLAIMER_EN = (
    "Based on published eligibility rules only; confirm with the implementing office. "
    "Not legal advice."
)
DISCLAIMER_ML = (
    "പ്രസിദ്ധീകരിച്ച യോഗ്യതാ നിബന്ധനകളെ അടിസ്ഥാനമാക്കി മാത്രം; "
    "നടപ്പാക്കുന്ന ഓഫീസുമായി സ്ഥിരീകരിക്കുക. നിയമ ഉപദേശമല്ല."
)


def _fmt_income(value: float | None) -> str:
    if value is None:
        return "n/a"
    return f"Rs.{int(value):,}"


def _rule_phrase_en(
    rule: str,
    scheme: dict[str, Any],
    profile: MatchProfile,
) -> str:
    rules = scheme.get("eligibility_rules") or {}
    if rule == "min_age":
        return f"age {profile.age} meets minimum age {rules.get('min_age')}"
    if rule == "max_age":
        return f"age {profile.age} within maximum age {rules.get('max_age')}"
    if rule == "max_annual_income":
        return (
            f"annual income {_fmt_income(profile.annual_income)} "
            f"within cap {_fmt_income(rules.get('max_annual_income'))}"
        )
    if rule == "max_monthly_household_income":
        return (
            f"monthly household income {_fmt_income(profile.monthly_household_income)} "
            f"within cap {_fmt_income(rules.get('max_monthly_household_income'))}"
        )
    if rule == "gender":
        return f"gender '{profile.gender}' matches required '{rules.get('gender')}'"
    if rule == "marital_status":
        return f"marital status '{profile.marital_status}' is in allowed list"
    if rule == "deserted_7_years_over_50":
        return f"deserted case with age {profile.age} (≥50 required)"
    if rule == "disability_required":
        return "disability requirement satisfied"
    if rule == "min_disability_percent":
        return (
            f"disability {profile.disability_percent}% meets minimum "
            f"{rules.get('min_disability_percent')}%"
        )
    if rule == "occupations":
        return "occupation intersects scheme occupation list"
    if rule == "categories":
        return "category intersects scheme category list"
    if rule == "states":
        return f"state '{profile.state or 'Kerala'}' is in scheme states"
    if rule == "land_ownership":
        return "land ownership requirement satisfied"
    if rule == "land_ownership_landless":
        return "profile indicates landless / no ownership"
    return rule.replace("_", " ")


def _rule_phrase_ml(
    rule: str,
    scheme: dict[str, Any],
    profile: MatchProfile,
) -> str:
    # Concise ML mirrors; native review pending for production.
    rules = scheme.get("eligibility_rules") or {}
    if rule == "min_age":
        return f"പ്രായം {profile.age} ഏറ്റവും കുറഞ്ഞ പ്രായം {rules.get('min_age')} നിറവേറ്റുന്നു"
    if rule == "max_annual_income":
        return (
            f"വാർഷിക വരുമാനം {_fmt_income(profile.annual_income)} "
            f"പരിധി {_fmt_income(rules.get('max_annual_income'))} യിൽ ഉൾപ്പെടുന്നു"
        )
    if rule == "max_monthly_household_income":
        return (
            f"മാസ വരുമാനം {_fmt_income(profile.monthly_household_income)} "
            f"പരിധിക്കുള്ളിൽ"
        )
    if rule == "gender":
        return f"ലിംഗം '{profile.gender}' യോജിക്കുന്നു"
    if rule == "marital_status":
        return f"വൈവാഹിക നില '{profile.marital_status}' അനുവദനീയമാണ്"
    if rule == "disability_required":
        return "ഭിന്നശേഷി നിബന്ധന നിറവേറ്റുന്നു"
    if rule == "min_disability_percent":
        return (
            f"ഭിന്നശേഷി {profile.disability_percent}% "
            f"കുറഞ്ഞത് {rules.get('min_disability_percent')}% നിറവേറ്റുന്നു"
        )
    if rule == "occupations":
        return "തൊഴിൽ പട്ടികയുമായി യോജിക്കുന്നു"
    if rule == "categories":
        return "വിഭാഗം പട്ടികയുമായി യോജിക്കുന്നു"
    if rule == "states":
        return f"സംസ്ഥാനം '{profile.state or 'Kerala'}' യോജിക്കുന്നു"
    if rule == "land_ownership":
        return "ഭൂമി ഉടമസ്ഥത നിബന്ധന നിറവേറ്റുന്നു"
    return rule.replace("_", " ") + " [ML]"


def template_explanation(
    *,
    scheme: dict[str, Any],
    profile: MatchProfile,
    matched_rules: list[str],
    unmatched_rules: list[str],
    missing_fields: list[str],
    status: str,
) -> LocalizedText:
    name_en = (scheme.get("scheme_name") or {}).get("en") or scheme.get("id", "scheme")
    name_ml = (scheme.get("scheme_name") or {}).get("ml") or name_en

    if matched_rules:
        parts_en = [_rule_phrase_en(r, scheme, profile) for r in matched_rules]
        parts_ml = [_rule_phrase_ml(r, scheme, profile) for r in matched_rules]
        body_en = "; ".join(parts_en) + "."
        body_ml = "; ".join(parts_ml) + "."
    else:
        body_en = "No structured hard-filter fields constrained this match."
        body_ml = "ഘടനാപരമായ കഠിന ഫിൽട്ടറുകൾ ഈ പൊരുത്തത്തെ നിയന്ത്രിച്ചില്ല."

    if status == "uncertain":
        prefix_en = f"Possibly relevant for {name_en} (needs verification). "
        prefix_ml = f"{name_ml} — സ്ഥിരീകരണം ആവശ്യം. "
    else:
        prefix_en = f"Likely eligible for {name_en}: "
        prefix_ml = f"{name_ml} യ്ക്ക് യോഗ്യതയുണ്ടാകാം: "

    extra_en = ""
    extra_ml = ""
    if missing_fields:
        extra_en += f" Missing profile fields: {', '.join(missing_fields)}."
        extra_ml += f" കുറവുള്ള ഫീൽഡുകൾ: {', '.join(missing_fields)}."
    if unmatched_rules:
        extra_en += f" Unmatched: {', '.join(unmatched_rules)}."
    verify = bool((scheme.get("eligibility_rules") or {}).get("verify"))
    if verify:
        notes = (scheme.get("eligibility_rules") or {}).get("verify_notes") or ""
        extra_en += " Official rules marked verify=true"
        if notes:
            extra_en += f" ({notes[:180]}{'…' if len(notes) > 180 else ''})"
        extra_en += "."
        extra_ml += " ഔദ്യോഗിക നിബന്ധനകൾ verify=true ആയി അടയാളപ്പെടുത്തിയിരിക്കുന്നു."

    en = prefix_en + body_en + extra_en + " " + DISCLAIMER_EN
    ml = prefix_ml + body_ml + extra_ml + " " + DISCLAIMER_ML
    return LocalizedText(en=en.strip(), ml=ml.strip())


def llm_explanation_stub(
    *,
    scheme: dict[str, Any],
    profile: MatchProfile,
    matched_rules: list[str],
    status: str,
) -> LocalizedText | None:
    """Optional LLM polish. Never called for eligibility decisions.

    Returns None when LLM_API_KEY is unset or call is unavailable — caller
    falls back to templates.
    """
    api_key = os.environ.get("LLM_API_KEY", "").strip()
    if not api_key:
        return None
    # Stub: do not invent thresholds; only rephrase matched_rules.
    # Real provider wiring is deferred; keep deterministic templates as source of truth.
    base = template_explanation(
        scheme=scheme,
        profile=profile,
        matched_rules=matched_rules,
        unmatched_rules=[],
        missing_fields=[],
        status=status,
    )
    base.en = "[LLM stub] " + base.en
    base.ml = "[LLM stub] " + base.ml
    return base


def build_explanation(
    *,
    scheme: dict[str, Any],
    profile: MatchProfile,
    matched_rules: list[str],
    unmatched_rules: list[str],
    missing_fields: list[str],
    status: str,
) -> LocalizedText:
    llm = llm_explanation_stub(
        scheme=scheme,
        profile=profile,
        matched_rules=matched_rules,
        status=status,
    )
    if llm is not None:
        return llm
    return template_explanation(
        scheme=scheme,
        profile=profile,
        matched_rules=matched_rules,
        unmatched_rules=unmatched_rules,
        missing_fields=missing_fields,
        status=status,
    )


def generator_mode() -> str:
    return "llm" if os.environ.get("LLM_API_KEY", "").strip() else "template"
